File: testing/hash_grouping/go/algo_group.go
package main

type Availability struct {
	Count     int
	Capacity  int
	Available int
}

type Location struct {
	Building string
	Room     string
}

type Instructor struct {
	Id   int
	Name string
}

type Crn struct {
	Crn         int
	Campus      string
	Section     string
	Instructors []Instructor
	Locations   []Location
	Enrollment  Availability
	Waitlist    Availability
}

type Classtime struct {
	Start_time int
	End_time   int
	Days       int
	Start_date int
	End_date   int
}

type CourseGroup struct {
	Hash       string
	Classtimes []Classtime
	Crns       []Crn
}

type Course struct {
	Name   string
	Title  string
	Groups []CourseGroup
}

// func peek[T any](stack []T) T {
// 	return stack[len(stack) - 1]
// }

func check_times(to_check []Classtime, schedule []Classtime) bool {
	for i := range to_check {
		classtime_a := to_check[i]
		for j := range schedule {
			classtime_b := schedule[j]
			if !(classtime_a.Start_time > classtime_b.End_time || classtime_b.Start_time > classtime_a.End_time){
				if (classtime_a.Days & classtime_b.Days) != 0 {
					if !(classtime_a.Start_date > classtime_b.End_date || classtime_b.Start_date > classtime_a.End_date){
						return false
					}
				}
			}
		}
	}
	return true
}

func CreateSchedules(courses []Course) [][]CourseGroup{
	first := courses[0].Groups[0]
	group_idxs, temp, temp_classtimes, results := []int{1}, []CourseGroup{first}, [][]Classtime{first.Classtimes}, [][]CourseGroup{}
	course_idx, group_idx := 0, -1

	for course_idx >= 0 {
		if course_idx+1 != len(temp) {
			group_idx = group_idxs[len(group_idxs)-1]
			group_idxs = group_idxs[:len(group_idxs)-1]
		} else {
			course_idx++
			group_idx = 0
		}

		current_course := courses[course_idx].Groups
		for group_idx < len(current_course) {
			group := current_course[group_idx]
			if len(temp) == 0 || len(group.Classtimes) == 0 || check_times(group.Classtimes, temp_classtimes[len(temp_classtimes)-1]) {
				temp = append(temp, group)
				if len(temp_classtimes) == 0 {
					temp_classtimes = append(temp_classtimes, group.Classtimes)
				} else {
					temp_classtimes = append(temp_classtimes, temp_classtimes[len(temp_classtimes)-1])
					temp_classtimes[len(temp_classtimes)-1] = append(temp_classtimes[len(temp_classtimes)-1], group.Classtimes...)
				}
				group_idxs = append(group_idxs, group_idx+1)
				break
			} else {
				group_idx++
			}
		}

		if len(temp) == len(courses) {
			results = append(results, temp)
			temp = temp[:len(temp)-1]
			temp_classtimes = temp_classtimes[:len(temp_classtimes)-1]
		} else if group_idx >= len(current_course) {
			if course_idx > 0 {
				temp = temp[:len(temp)-1]
				temp_classtimes = temp_classtimes[:len(temp_classtimes)-1]
			}
			course_idx--
		} else {
		}
	}
	return results
}

File: testing/hash_grouping/cs/AlgoGroup.cs
using System.Collections.Generic;
using System.Linq;

namespace Scheduler
{
    public class AlgoGroup
    {
        private static bool CheckTimes(IEnumerable<Classtime> toCheck, IEnumerable<Classtime> schedule)
        {
            foreach (var classtimeA in toCheck)
            {
                foreach (var classtimeB in schedule)
                {
                    if (!(classtimeA.Start_Time > classtimeB.End_Time || classtimeB.Start_Time > classtimeA.End_Time))
                    {
                        if ((classtimeA.Days & classtimeB.Days) != 0)
                        {
                            if (!(classtimeA.Start_Date > classtimeB.End_Date || classtimeB.Start_Date > classtimeA.End_Date))
                            {
                                return false;
                            }
                        }
                    }
                }
            }
            return true;
        }
        public static IEnumerable<IEnumerable<CourseGroup>> CreateSchedules(IEnumerable<Course> courses)
        {
            courses = courses.OrderBy(course => course.Groups.Count());
            var first = courses.First().Groups.First();
            var groupIdxs = new Stack<int>(new[] { 1 });
            var temp = new Stack<CourseGroup>(new[] { first });
            var tempClasstimes = new Stack<IEnumerable<Classtime>>(new[] { first.Classtimes });
            var results = new List<IEnumerable<CourseGroup>>();
            var courseIdx = 0;
            var groupIdx = -1;

            while (courseIdx >= 0)
            {
                if (courseIdx + 1 != temp.Count)
                {
                    groupIdx = groupIdxs.Pop();
                }
                else
                {
                    courseIdx++;
                    groupIdx = 0;
                }

                var currentCourse = courses.ElementAt(courseIdx).Groups;
                while (groupIdx < currentCourse.Count())
                {
                    var group = currentCourse.ElementAt(groupIdx);
                    if (temp.Count == 0 || group.Classtimes.Count() == 0 || CheckTimes(group.Classtimes, tempClasstimes.Peek()))
                    {
                        temp.Push(group);
                        if (tempClasstimes.Count == 0)
                        {
                            tempClasstimes.Push(group.Classtimes);
                        }
                        else
                        {
                            tempClasstimes.Push(tempClasstimes.Peek().Concat(group.Classtimes));
                        }
                        groupIdxs.Push(groupIdx + 1);
                        break;
                    }
                    else
                    {
                        groupIdx++;
                    }
                }

                if (temp.Count == courses.Count())
                {
                    results.Add(temp.ToList());
                    temp.Pop();
                    tempClasstimes.Pop();
                }
                else if (groupIdx >= currentCourse.Count())
                {
                    if (courseIdx > 0)
                    {
                        temp.Pop();
                        tempClasstimes.Pop();
                    }
                    courseIdx--;
                }
                else { }
            }

            return results;
        }
    }
}
